### DISCUSSION TITLE
SELECT_NEXT_PATTERN_CC_ABSOLUTE

### DISCUSSION THREAD

**Discussion by sirjohndfox** [hydrogen-music/hydrogen#1631] [Hydrogen on Linux]
Gday,

I'm running 1.1.1-'7e49a4fa'.   I cant seem to get this midi action to work.  Do I understand it's function correctly?

Say if I set it to CC 116 -

A value of 0 should select pattern 0
A value of 1 should select pattern 1 etc.

Is there anyway I could get it to function like select and play next pattern?

I've tried to dig around in the core but I cant really work out what's going on.  Some pointers of where to start looking would be appreciated.
Cheers.

**Comment by elpescado** [comment]
If you need to move to pattern relative to selected one, you should use `SELECT_NEXT_PATTERN_CC_RELATIVE` MIDI action. I.e. if currently selected pattern is pattern number 4, and you want to jump to pattern 6, you can trigger `SELECT_NEXT_PATTERN_CC_RELATIVE` with value 2.

**Comment by elpescado** [comment]
BTW. Most MIDI actions in Hydrogen come in "absolute" and "relative" variants. "Absolute" ones set destination value to value specified in CC message (eg. "set master volume to 42"), whereas "relative" ones adjust current values by CC message value ("increase master volume by 3").

**Comment by sirjohndfox** [comment]
Ah right I just realised that SELECT_NEXT_PATTERN_CC_ABSOLUTE works as expected in *single pattern mode*.

What I'm after is really SELECT_ONLY_NEXT_PATTERN in stacked mode but with the pattern number coming from the CC value.  I've written a bash script to populate the midi settings with separate controllers for an arbitrary number of patterns but it's quite messy.

It's a nice function to be able to trigger the next pattern to play at the end of the currently playing pattern.   The function I want should be very easy to patch - I just don't have much c++ experience and can't yet grok the code.

I take it a vector is some kind of array but I can't see where the values are being passed.

Thanks for your help anyway!

Cheers
John


Sent with Proton Mail secure email.

------- Original Message -------

**Comment by theGreatWhiteShark** [comment]
Hey @sirjohndfox ,

Sounds like a very reasonable feature (especially since UX of stacked pattern mode increased a lot lately). Probably this part of MIDI actions grew naturally and the CC version for stacked mode was somehow missed.

I introduced a `SELECT_ONLY_NEXT_PATTERN_CC_ABSOLUTE` MIDI action in #1640. Could you check whether this works for you?

**Comment by sirjohndfox** [comment]
Gday @theGreatWhiteShark 
thanks heaps for having a shot at this feature.
I'm not really sure how to test a pull request that hasn't been merged.  I need to learn git one of these days.

I did:
git  clone https://github.com/hydrogen-music/hydrogen
cd hydrogen
git fetch origin pull/1640/head:test-branch
git checkout test-branch

then I tried to build as normal but i got some errors:


hydrogen/src/gui/src/Reporter.cpp:179:23: error: comparison between distinct pointer types ‘QPushButton*’ and ‘QAbstractButton*’ lacks a cast
    if ( pLogButton == pPushed ) {
                       ^~~~~~~
/home/john/SOFTWARE/hydrogen/src/gui/src/Reporter.cpp:182:27: error: comparison between distinct pointer types ‘QAbstractButton*’ and ‘QPushButton*’ lacks a cast
    } else if ( pPushed == pIssuesButton ) {
                           ^~~~~~~~~~~~~

Am I doing something wrong?  Maybe my qt libraries are a bit old or something?

Thanks for your help.
Cheers
John

**Reply by theGreatWhiteShark** [reply]
> I'm not really sure how to test a pull request that hasn't been merged.

You did perfectly fine. But your compiler throws an error at a problem ours as well as the one used in out automated unit tests does not complain about.

Could you fetch again? Should be fixed now.

**Reply by sirjohndfox** [reply]
alas no dice
`
/home/john/SOFTWARE/hydrogen/src/gui/src/Reporter.cpp: In member function ‘void Reporter::on_finished(int, QProcess::ExitStatus)’:
/home/john/SOFTWARE/hydrogen/src/gui/src/Reporter.cpp:179:49: error: invalid static_cast from type ‘QPushButton*’ to type ‘QAbstractButton*’
    if ( static_cast<QAbstractButton*>(pLogButton) == pPushed ) {
                                                 ^
/home/john/SOFTWARE/hydrogen/src/gui/src/Reporter.cpp:183:57: error: invalid static_cast from type ‘QPushButton*’ to type ‘QAbstractButton*’
    else if ( static_cast<QAbstractButton*>(pIssuesButton) == pPushed ) {`

**Reply by sirjohndfox** [reply]
@theGreatWhiteShark 

My fault I'm using an old debian stable dist with gcc 8.  It compiled fine on a newer system with gcc 12.  It'll take a little while to test though, cause the newer machine is in another room and I'm having a Sunday lie in :)   I'll let you know how it goes.

**Reply by sirjohndfox** [reply]
That said I should be able to compile on my stable debian.  The error is in a weird little corner of the code..
The last version I compiled on that system was hydrogen 1.1.1-'7e49a4fa'.  and that source doesn't have Reporter.cpp in it.  Is it a dialogue to send a bug report?  Perhaps I should open a new issue?

**Reply by theGreatWhiteShark** [reply]
> That said I should be able to compile on my stable debian

Definitely. Even way older versions should compile as we explicitly want to support old hardware.
